homeassistant/components/proximity/coordinator.py

```diff
@@ -69,189 +82,227 @@ def __init__(
-
-        zone_state = self.hass.states.get(f"zone.{self.proximity_zone}")
-        proximity_latitude = (
-            zone_state.attributes.get(ATTR_LATITUDE) if zone_state else None
-        )
-        proximity_longitude = (
-            zone_state.attributes.get(ATTR_LONGITUDE) if zone_state else None
+    def _convert(self, value: float | str) -> float | str:
+        """Round and convert given distance value."""
+        if isinstance(value, str):
+            return value
+        return round(
+            DistanceConverter.convert(
+                value,
+                UnitOfLength.METERS,
+                self.unit_of_measurement,
+            )
         )
 
-        # Check for devices in the monitored zone.
-        for device in self.proximity_devices:
-            if (device_state := self.hass.states.get(device)) is None:
-                devices_to_calculate = True
-                continue
-
-            if device_state.state not in self.ignored_zones:
-                devices_to_calculate = True
-
-            # Check the location of all devices.
-            if (device_state.state).lower() == (self.proximity_zone).lower():
-                device_friendly = device_state.name
-                devices_in_zone.append(device_friendly)
-
-        # No-one to track so reset the entity.
-        if not devices_to_calculate:
-            _LOGGER.debug("no devices_to_calculate -> abort")
-            return {
-                "dist_to_zone": DEFAULT_DIST_TO_ZONE,
-                "dir_of_travel": DEFAULT_DIR_OF_TRAVEL,
-                "nearest": DEFAULT_NEAREST,
-            }
-
-        # At least one device is in the monitored zone so update the entity.
-        if devices_in_zone:
-            _LOGGER.debug("at least one device is in zone -> arrived")
-            return {
-                "dist_to_zone": 0,
-                "dir_of_travel": "arrived",
-                "nearest": ", ".join(devices_in_zone),
-            }
-
-        # We can't check proximity because latitude and longitude don't exist.
-        if "latitude" not in state_change_data.new_state.attributes:
-            _LOGGER.debug("no latitude and longitude -> reset")
-            return self.data
-
-        # Collect distances to the zone for all devices.
-        distances_to_zone: dict[str, float] = {}
-        for device in self.proximity_devices:
-            # Ignore devices in an ignored zone.
-            device_state = self.hass.states.get(device)
-            if not device_state or device_state.state in self.ignored_zones:
-                continue
+    def _calc_distance_to_zone(
+        self,
+        zone: State,
+        device: State,
+        latitude: float | None,
+        longitude: float | None,
+    ) -> float | None:
+        if device.state.lower() == self.proximity_zone.lower():
+            _LOGGER.debug(
+                "%s: %s in zone -> distance=0", self.friendly_name, device.entity_id
+            )
+            return 0
```

Comment:
I think it is wrong to set distance to 0 in case the tracked entity is in the monitored zone.
You can be within the monitored zone but still 1 or 2km away (_depending on the radius of the monitored zone_)
Therefore I think the condition that a tracked entity is in the monitored zone, should only influence the "direction of travel" and set this to "arrived".
With this the user can still easily figure out if some tracked entity is in the monitored zone without the need of any helper entities

Reply:
c61ccc7